src/Functions/toBool.cpp

```diff
@@ -54,7 +56,10 @@ namespace
                 }
             };
 
-            auto func_cast = createInternalCast(arguments[0], result_type, CastType::nonAccurate, {}, nullptr);
+            const bool needs_accurate_or_null
+                = arguments[0].type->isNullable() || arguments[0].type->lowCardinality() || arguments[0].type->isLowCardinalityNullable();
+            auto cast_type = needs_accurate_or_null ? CastType::accurateOrNull : CastType::nonAccurate;
+            auto func_cast = createInternalCast(arguments[0], result_type, cast_type, {}, context);
```

Comment:
I merged master, but this (`context`) needs adjusting - but I will not be able to do it till mid January.